### PR TITLE
Tango

### PR DESCRIPTION
**Pull request by dirkgr**
1700 lines is pretty uncomfortable to review, so let me attempt a guide:

Steps
====

The most important thing is the `Step` class. It defines one step in a workflow. Users are expected to just write a `run()` method. The `run()` method must have parameters with type hints. `from_params()` reads those type hints to construct `Step`s. If the `run()` method takes a parameter of type `T`, then `from_params()` assumes the constructor of that step takes a `Union[T, Step[T]]`. In other words, you can provide the `T` directly, or you can put in a `Step` that outputs a `T`. Making `Step`s the input is how you define a DAG of tasks. The `Step` code makes sure to replace inputs of type `Step` with the `Step`'s results before the `run()` method runs.

Hopefully most of this magic will remain hidden to users, but as a reviewer you have to understand it.

`Step`s also store some settings:
 * `DETERMINSTIC` describes whether this step can be relied upon to produce the same results every time when given the same inputs. If this is `False`, the step can't be cached, and neither can any step that depends on it.
 * `CACHEABLE` provides a direct way to turn off caching. For example, a step that reads a HuggingFace dataset doesn't need to be cached, because HuggingFace datasets already have their own caching mechanism. But it's still a deterministic step, and all following steps are allowed to cache.
 * `VERSION` is optional, but recommended. This gives the user a way to tell Tango that a step has changed during development, and should now be recomputed. This doesn't invalidate the old results, so when you revert your code, the old cache entries will stick around and be picked up.
 * `FORMAT`: See below

Those settings above are per `Step` class. Every *instance* of `Step` has some more settings. You can override the format and whether to cache the results, and there are two more:
 * `step_name` allows you to give your step a useful name that you can use to refer to it. This name will be used in many places, like names of directories for results, log messages, etc. If this is not given, the step's unique id stands in.
 * `produce_results` specifies whether this is a step where we care about the the results. For example, you might build a long pipeline of steps, but you really only want to look at the evaluation at the end. In that case, the evaluation step is the only step where you would set `produce_results` to `True`. If none of your steps have `produce_results` set to `True`, Tango will do nothing. It will only run steps that are necessary to produce the results you need.

 Every step has a unique id of the form `f"{self.__class__.__name__}-{self.VERSION}-{hash of input}"`. 

`RefStep`
-----
`_RefStep` is a fake stand-in for a real step. This is used when parsing a DAG of steps. Every step gets parsed on its own, and references to other steps are parsed as a `_RefStep`. Then later, when all the steps are known, `_RefStep`s get resolved to real steps. A final DAG should never contain a `_RefStep`.

Formats
====
The `Format` classes are basically mappings from Python objects to disk and back. They have a `read()` and a `write()` method. They read and write directories, not files. At the end of the `write()` method, the object must be completely written, but the `read()` method can return an object that will do the actual reading lazily. We might use this for reading datasets, where a `read()` method can return immediately, and return an object that reads actual instances later.

Every `Step` defines the `Format` it wants to use to serialize and deserialize its result.

`DillFormat` is the default format for all steps. It uses `dill` (a better version of `pickle`) to serialize and deserialize objects. It is surprisingly flexible and can handle almost all Python objects (including (some) functions).

`StepCache`
===
`StepCache` is a mapping from instances of `Step` to the results of that step. There are two implementations, `MemoryStepCache` and `DirectoryStepCache`. `DirectoryStepCache` is the component that uses the `Format` classes to cache outputs in a directory. When you run `allennlp tango -s serialization_dir`, then the step cache ends up in `serialization_dir/step_cache`.

Dataset
====
Tango datasets are a bit different, in that they contain all splits simultaneously, as well as a vocabulary. The definition of `AllenNlpDataset` is this:
```Python
@dataclass
class AllenNlpDataset:
    splits: Mapping[str, Sequence[Any]]
    vocab: Optional[Vocabulary] = None
    metadata: Mapping[str, Any] = field(default_factory=dict)
```
For bigger datasets, the idea is that we write something that returns an object which can read instances lazily. For example, You could imagine that the `Sequence[Any]` that contains the instances refers to a directory on disk somewhere, and reads one file per instance. But that's all future stuff. In this PR, all the `Sequence`s are `List`s.

One important thing is this: In Tango, *all* datasets are map-style datasets (to use PyTorch terminology). Iterator-style datasets ("lazy datasets", as we call them in AllenNLP) are a pain, and are not necessary given the right tooling.

Dataloader
====
`TangoDataLoader` is simpler than the original AllenNLP `DataLoader`. Original data loaders are responsible for a ton of stuff, and have a bit of a messy API owing to the history of their development. Tango data loaders can be simpler because they don't have to deal with lazy datasets or multiprocessing. All they do is make batches out of `Sequence[Instance]`.

The old `MultiprocessingDataLoader` handles a lot of scenarios:
 * normal, equal-sized batches
 * shuffling
 * manually setting the number of batches per epoch
 * using a PyTorch sampler

I have split these out into separate data loader classes, which are composable (i..e, one data loader feeds into another). If you think composing data loaders is too complicated for researchers who'd rather not think about it, I'm open to that argument. I think I'd want to keep this API internally because it makes for small classes that are easy to understand, but maybe we write an interface that makes it easier to configure these.

Example Steps
====

`TrainingStep`
----
This step takes basically all the inputs that `GradientDescentTrainer` needs, plus a dataset and split name, and trains a model on it. It adapters to `GradientDescentTrainer`. In another iteration, I want to switch to getting rid of the adapter and having the training code directly in the step, since I am pretty unhappy with the inconsistent trainer API we have.

`EvaluationStep`
----
This step takes a model, a dataset, and the name of a split, and produces an evaluation result. It's roughly equivalent to `allennlp evaluate`.

`HuggingfaceDataset`
----
This step loads a HuggingFace dataset and puts it into AllenNLP dataset format. It's a one-liner of a step.

`HuggingfaceTokenize`
----
This step uses the HuggingFace tokenizers to turn every `str` in a given dataset into a `TransformerTextField`. That sounds like it would be super useful, but I actually ended up not using it for PIQA. So far I have found that I always need some more specialized treatment of the dataset before turning it into the input to a model.

### CODE REVIEW

**Comment by dirkgr** [comment]
I refactored the code a bunch to be mergeable, but the tests are not refactored yet, and I haven't written documentation. I want to add those before making this into a non-draft PR.

**Comment by AkshitaB** [comment]
@dirkgr I haven't looked at the code in detail yet. Had a question about this:

```produce_results specifies whether this is a step where we care about the the results. For example, you might build a long pipeline of steps, but you really only want to look at the evaluation at the end. In that case, the evaluation step is the only step where you would set produce_results to True. If none of your steps have produce_results set to True, Tango will do nothing. It will only run steps that are necessary to produce the results you need.```

So, if you have :
1. Preprocess data and dataloader
2. TrainingStep
3. EvalStep

And you only set `produce_results=True` for the EvalStep, it will not save the model in the training step?

**Comment by dirkgr** [comment]
It will save every step in between that's cacheable. If that's not clear, I need to spruce up the docs.

**Comment by AkshitaB** [comment]
> It will save every step in between that's cacheable. If that's not clear, I need to spruce up the docs.

I guess the distinction between producing results and caching the outputs of the step isn't clear to me.

**Comment by dirkgr** [comment]
Maybe it should always produce all results. Or we'll invert the switch. We could call it `only_if_needed`, which is almost exactly the opposite of `produce_results`.

It is sometimes convenient to define a big experiment, but only run some parts of it. That's what `produce_results` is for. Explaining it better in the docs is one thing, but making it clear in the code is much better.

**Comment by dirkgr** [comment]
@AkshitaB, I made that change. I hope it is clearer now. Users can completely forget about this setting now.

File: allennlp/tango/dataset.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    This step creates an `AllenNlpDataset` from old-school dataset readers. If you're
+    tempted to write a new `DatasetReader`, and then use this step with it, don't.
+    Just write a `Step` that creates the `AllenNlpDataset` you need directly.
+    """
+
+    DETERMINISTIC = True  # We're giving the dataset readers some credit here.
+    CACHEABLE = True
+    VERSION = "002"
+
+    def run(self, reader: DatasetReader, splits: Dict[str, str]) -> AllenNlpDataset:  # type: ignore
```
Comment (dirkgr) [inline allennlp/tango/dataset.py:40]: Look into whether `hash_object(DatasetReader)` does what you think it does.

Comment (dirkgr) [reply allennlp/tango/dataset.py:40]: Make it possible to override how objects are hashed.

File: allennlp/tango/step.py [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+            return [cls._replace_steps_with_results(i, cache) for i in o]
+        elif isinstance(o, Set):
+            return {cls._replace_steps_with_results(i, cache) for i in o}
+        elif isinstance(o, Dict):
+            return {key: cls._replace_steps_with_results(value, cache) for key, value in o.items()}
+        else:
+            return o
+
+    def result(self, cache: Optional[StepCache] = None) -> T:
+        """Returns the result of this step. If the results are cached, it returns those. Otherwise it
```
Comment (dirkgr) [inline allennlp/tango/step.py:478]: Set random seeds before we call `run()`

Comment (dirkgr) [reply allennlp/tango/step.py:478]: Make it depend on `self.DETERMINISTIC`?

File: allennlp/tango/step.py [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+        else:
+            self.temp_dir_for_run = step_dir / "run"
+            try:
+                self.temp_dir_for_run.mkdir(exist_ok=True, parents=True)
+                return self.run(**kwargs)
+            finally:
+                # No cleanup, as we want to keep the directory for restarts or serialization.
+                self.temp_dir_for_run = None
+
+    def temp_dir(self) -> PathLike:
```
Comment (dirkgr) [inline allennlp/tango/step.py:458]: Call this `run_dir` or something? It's not truly temporary.

**Review by dirkgr — COMMENTED** [review]
Find a way to annotate experimental APIs and mark all of Tango with it.

```
@experimental
def foo():
    pass

@experimental
class Foo:
    pass
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -58,6 +58,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added `TransformerModule._post_load_pretrained_state_dict_hook()` method. Can be used to modify `missing_keys` and `unexpected_keys` after
   loading a pretrained state dictionary. This is useful when tying weights, for example.
 - Added an end-to-end test for the Transformer Toolkit.
+- Added Tango components, to be explored in detail in a later post.
```
Comment (epwalsh) [inline CHANGELOG.md:61]: This needs to be moved up to the "Unreleased" section

File: setup.py [hunk 5]
```diff
@@ -76,6 +76,9 @@
         "checklist==0.0.11",
         "wandb>=0.10.0,<0.12.0",
         "huggingface_hub>=0.0.8",
+        "datasets>=1.2.1",
```
Comment (epwalsh) [inline setup.py:79]: We should probably pin this tighter:
```suggestion
        "datasets>=1.2.1,<1.3.0",
```

File: allennlp/tango/step.py (hunk 2 above)
Comment (dirkgr) [reply allennlp/tango/step.py:496]: Fixed

File: allennlp/tango/step.py (hunk 3 above)
Comment (dirkgr) [reply allennlp/tango/step.py:476]: It's now called `work_dir`.

File: allennlp/tango/dataloader.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    def __iter__(self) -> Iterator[TensorDict]:
+        instances: Sequence[Instance]
+        if self.shuffle:
+            instances_as_list_just_to_make_mypy_happy = list(
+                self.instances
+            )  # make a new list pointing to the same instance objects
+            random.shuffle(instances_as_list_just_to_make_mypy_happy)
+            instances = instances_as_list_just_to_make_mypy_happy
+        else:
+            instances = self.instances
```
Comment (epwalsh) [inline allennlp/tango/dataloader.py:119]: It might be really expensive to create a list here if your actual dataset is backed by something like an Arrow table. Alternatively, you could just shuffle a list of indices, and then iterate over the list of indices.

**Review by epwalsh — CHANGES_REQUESTED** [review]
A few more comments

File: CHANGELOG.md (hunk 4 above)
Comment (dirkgr) [reply CHANGELOG.md]: Indeed

File: allennlp/tango/evaluation.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        generator_tqdm = Tqdm.tqdm(iter(concrete_data_loader))
+
+        # Number of batches in instances.
+        batch_results = []
+        # Number of batches where the model produces a loss.
+        loss_count = 0
+        # Cumulative loss
+        total_loss = 0.0
+
+        with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp/tango/evaluation.py:73]: @epwalsh, what's the better context manager?

File: setup.py (hunk 5 above)
Comment (dirkgr) [reply setup.py]: Why are we keeping it so small? Latest one is `1.10.2`, no?

File: allennlp/tango/dataset.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+from dataclasses import dataclass, field
+from typing import Mapping, Any, Optional, Sequence, Dict
+
+from allennlp.data import Vocabulary, DatasetReader, Instance
+from allennlp.tango.step import Step
+from tqdm import tqdm
+
+
+@dataclass
+class AllenNlpDataset:
```
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp/tango/dataset.py]: Renamed

File: allennlp/tango/evaluation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp/tango/evaluation.py:74]: It's called `torch.inference_mode()` 😱

File: setup.py (hunk 5 above)
Comment (epwalsh) [reply setup.py]: Oh, yea. So how about `<2.0`?

File: allennlp/tango/dataset.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+from dataclasses import dataclass, field
+from typing import Mapping, Any, Optional, Sequence, Dict
+
+from allennlp.data import Vocabulary, DatasetReader, Instance
+from allennlp.tango.step import Step
+from tqdm import tqdm
+
+
+@dataclass
+class DatasetDict:
```
Comment (epwalsh) [inline allennlp/tango/dataset.py:16]: If we're going to call this `DatasetDict` (which I like), I think this should act like a `dict`. At least we should implement `__getitem__` so you can do `dataset_dict[split_name]`.

File: setup.py (hunk 5 above)
Comment (dirkgr) [reply setup.py]: done

File: allennlp/tango/dataloader.py (hunk 6 above)
Comment (dirkgr) [reply allennlp/tango/dataloader.py]: Introducing `ShuffledSequence` 71411e15b9ab206960bffe4aeac624ec953bc31c

File: allennlp/tango/dataset.py (hunk 9 above)
Comment (dirkgr) [reply allennlp/tango/dataset.py:16]: I put in `__getitem__` and `__len__`, but I am actually not a big fan. I want people to be at least aware that those other members are there, and making it act more and more like a dict is putting in a lot of magic that might bite us later.

File: allennlp/tango/dataset.py (hunk 1 above)
Comment (dirkgr) [reply allennlp/tango/dataset.py]: Done

**Review by epwalsh — APPROVED** [review]
Thanks for addressing all my comments! LGTM ✅

File: allennlp/common/file_utils.py [hunk 10]
```diff
@@ -617,64 +615,91 @@ def __init__(
...
+                max_readers=cpu_count * 4,
+                max_spare_txns=cpu_count * 4,
+                metasync=False,
+                sync=True,
+                readahead=False,
+                meminit=False,
+                readonly=read_only,
+                lock=use_lock,
+            )
+            _active_tensor_caches[self.lmdb_env.path()] = self
```
Comment (AkshitaB) [inline allennlp/common/file_utils.py:694]: Where do we remove entries from `_active_tensor_caches`? I see it in an older commit, but not in the latest one.

**Review by AkshitaB — CHANGES_REQUESTED** [review]

File: allennlp/common/file_utils.py [hunk 11]
```diff
@@ -577,6 +580,17 @@ class TensorCache(MutableMapping[str, Tensor], ABC):
     runs at the same time.
     """
 
+    def __new__(cls, filename: Union[str, PathLike], *, read_only: bool = False, **kwargs):
+        # This mechanism makes sure we re-use open lmdb file handles. Lmdb has a problem when the same file is
+        # opened by the same process multiple times. This is our workaround.
+        filename = str(filename)
```
Comment (epwalsh) [inline allennlp/common/file_utils.py:586]: We should probably normalize `filename` to an absolute path here, right?

File: allennlp/common/file_utils.py (hunk 10 above)
Comment (dirkgr) [reply allennlp/common/file_utils.py:694]: `_active_tensor_caches` is a `WeakValueDict`, which removes entries automatically when the values are GCd.

**Review by AkshitaB — APPROVED** [review]

File: allennlp/common/file_utils.py (hunk 11 above)
Comment (dirkgr) [reply allennlp/common/file_utils.py:592]: I'll do you one better and do it by inode: 153bade2b61abb54c17807934e59ee11cfaf4957

That way even symlinks and hard links work correctly.

Comment (epwalsh) [reply allennlp/common/file_utils.py:592]: 💯
